pca.go

```diff
@@ -0,0 +1,77 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix"
+	"github.com/gonum/matrix/mat64"
+)
+
+// PrincipalComponents returns the k principal component direction vectors of
+// the matrix a, projected score vectors and variance values by singular value
+// decomposition of the input. The input a is a matrix where each observation
+// is a row and each column represents a variable.
+//
+// The slice weights is used to weight the observations. If weights is nil,
+// each weight is considered to have a value of one, otherwise the length of
+// weight must match the number of observations or PrincipalComponents will panic.
+//
+// If no singular value decomposition is possible, pc, proj and vars are returned
+// nil and ok is returned false. PrincipalComponents will panic if k is greater
+// than or equal to the number of rows, or if k is greater than the number of
+// columns.
+func PrincipalComponents(a mat64.Matrix, weights []float64, k int) (pc, proj *mat64.Dense, vars []float64, ok bool) {
```

Comment:
The reason I included it is that the projection is done on the centred and weighted data, which we already have. Otherwise the client needs to construct that again to get the projection.